src/gui/qgsmapcanvas.cpp

```diff
@@ -778,8 +796,26 @@ QgsRectangle QgsMapCanvas::imageRect( const QImage& img, const QgsMapSettings& m
 
 void QgsMapCanvas::mapUpdateTimeout()
 {
-  const QImage& img = mJob->renderedImage();
-  mMap->setContent( img, imageRect( img, mJob->mapSettings() ) );
+  QImage img = mJob->renderedImage();
+  magnifyImage( img );
+
+  mMap->setContent( img, imageRect( img, mSettings ) );
+}
+
+void QgsMapCanvas::magnifyImage( QImage &img )
+{
+  if ( mMagnificationLevel != 1.0 )
+  {
+    img = img.scaled( img.width() * mMagnificationLevel,
```

Comment:
@pblottiere that may be due to how the pal labeling engine is non-deterministic. Ie with complex labeling projects refreshing multiple times will lead to different placements.

I think this could be partially improved by adding a new "viewport extent" option to QgsMapSettings. If set, this new extent would be used instead of the map extent for the feature request rectangle. (Ie, allow you to render a larger extent then is actually visible)

 At the moment the DPI approach requests a smaller extent, so features will be missing from the render if they fall off the edges due to magnification. This may result in less features being passed to the label engine, leading to the different placements. You'll also notice that zooming in using magnifier will drop off labels on the edges which should be visible.

By adding a viewport extent we could still request and render all features from the full original extent as though it were visible, without altering the map scale and map to pixel, etc.